### ISSUE TITLE
Typesense crashes when locale of the field is ja (Japanese) in Ubuntu 22.04 ARM64 Docker

### ISSUE THREAD

**Issue by despenser08** [typesense/typesense#1002]
## Description

Typesense crashes if client tries to import Japanese in `locale: "ja"` field.

## Steps to reproduce

1. Create a collection that contains `locale: "ja"` field. I used the code using Node.js with `typesense@1.5.3`:

```ts
await typesense.collections().create({ name: "test", fields: [{ name: "japanese", type: "string", locale: "ja" }] });
```

2. Then, import the document to the collection created.

```ts
await typesense
      .collections("test")
      .documents()
      .import([{ japanese: "こんにちは" }, { japanese: "またね" }], { action: "upsert" });
```

## Expected Behavior

Typesense should not crash when importing documents in Japanese.

## Actual Behavior

```
E20230506 20:16:24.619741    12 backward.hpp:4199] Stack trace (most recent call last) in thread 12:
E20230506 20:16:24.619771    12 backward.hpp:4199] #16   Object "/usr/lib/aarch64-linux-gnu/ld-linux-aarch64.so.1", at 0xffffffffffffffff, in 
E20230506 20:16:24.619776    12 backward.hpp:4199] #15   Object "/usr/lib/aarch64-linux-gnu/libc.so.6", at 0xffffa5e75d1b, in 
E20230506 20:16:24.619777    12 backward.hpp:4199] #14   Object "/usr/lib/aarch64-linux-gnu/libc.so.6", at 0xffffa5e0d5c7, in 
E20230506 20:16:24.619781    12 backward.hpp:4199] #13   Object "/opt/typesense-server", at 0x1367cbb, in execute_native_thread_routine
E20230506 20:16:24.619782    12 backward.hpp:4199] #12 | Source "/usr/local/include/c++/10.3.0/thread", line 215, in operator()
E20230506 20:16:24.619784    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/thread", line 271, in _M_invoke<0>
E20230506 20:16:24.619791    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/thread", line 264, in __invoke<ThreadPool::ThreadPool(size_t)::<lambda()> >
E20230506 20:16:24.619793    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 95, in __invoke_impl<void, ThreadPool::ThreadPool(size_t)::<lambda()> >
E20230506 20:16:24.619796    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 60, in operator()
E20230506 20:16:24.619799    12 backward.hpp:4199]     | Source "/typesense/include/threadpool.h", line 57, in operator()
E20230506 20:16:24.619801    12 backward.hpp:4199]       Source "/usr/local/include/c++/10.3.0/future", line 1568, in _M_run [0x49c403]
E20230506 20:16:24.619803    12 backward.hpp:4199] #11 | Source "/usr/local/include/c++/10.3.0/future", line 1435, in _M_set_result
E20230506 20:16:24.619807    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/future", line 412, in call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>()>*, bool*), std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>()>*, bool*>
E20230506 20:16:24.619808    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/mutex", line 729, in __gthread_once
E20230506 20:16:24.619810    12 backward.hpp:4199]       Source "/usr/local/include/c++/10.3.0/aarch64-linux-gnu/bits/gthr-default.h", line 700, in _M_run [0x574a1b]
E20230506 20:16:24.619812    12 backward.hpp:4199] #10   Object "/usr/lib/aarch64-linux-gnu/libc.so.6", at 0xffffa5e125d3, in 
E20230506 20:16:24.619815    12 backward.hpp:4199] #9  | Source "/usr/local/include/c++/10.3.0/future", line 572, in operator()
E20230506 20:16:24.619817    12 backward.hpp:4199]       Source "/usr/local/include/c++/10.3.0/bits/std_function.h", line 622, in _M_do_set [0x499f9b]
E20230506 20:16:24.619820    12 backward.hpp:4199] #8  | Source "/usr/local/include/c++/10.3.0/bits/std_function.h", line 292, in __invoke_r<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<_Fn, _Alloc, _Res(_Args ...)>::_M_run<std::_Bind<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>()>, std::allocator<int>, void, {}>::<lambda()>, void>&>
E20230506 20:16:24.619823    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 115, in __invoke_impl<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<_Fn, _Alloc, _Res(_Args ...)>::_M_run<std::_Bind<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>()>, std::allocator<int>, void, {}>::<lambda()>, void>&>
E20230506 20:16:24.619827    12 backward.hpp:4199]       Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 60, in _M_invoke [0x5a4147]
E20230506 20:16:24.619828    12 backward.hpp:4199] #7  | Source "/usr/local/include/c++/10.3.0/future", line 1373, in operator()
E20230506 20:16:24.619830    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/future", line 1432, in __invoke_r<void, std::_Bind<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>()>&>
E20230506 20:16:24.620000    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 110, in __invoke_impl<void, std::_Bind<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>()>&>
E20230506 20:16:24.620012    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 60, in operator()<>
E20230506 20:16:24.620016    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/functional", line 499, in __call<void>
E20230506 20:16:24.620018    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/functional", line 416, in __invoke<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>&>
E20230506 20:16:24.620021    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 95, in __invoke_impl<void, Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>&>
E20230506 20:16:24.620023    12 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 60, in operator()
E20230506 20:16:24.620025    12 backward.hpp:4199]       Source "/typesense/src/index.cpp", line 699, in operator() [0x5a4043]
E20230506 20:16:24.620028    12 backward.hpp:4199] #6    Source "/typesense/src/index.cpp", line 635, in validate_and_preprocess [0x5a383f]
E20230506 20:16:24.620029    12 backward.hpp:4199] #5    Source "/typesense/src/index.cpp", line 312, in compute_token_offsets_facets [0x5a2ce3]
E20230506 20:16:24.620033    12 backward.hpp:4199] #4    Source "/typesense/src/index.cpp", line 1153, in tokenize_string_with_facets [0x5a1f33]
E20230506 20:16:24.620034    12 backward.hpp:4199] #3    Source "/typesense/src/tokenizer.cpp", line 30, in Tokenizer [0x63726b]
E20230506 20:16:24.620036    12 backward.hpp:4199] #2    Source "/typesense/src/tokenizer.cpp", line 63, in init [0x636f0f]
E20230506 20:16:24.620039    12 backward.hpp:4199] #1    Source "/typesense/external-Linux/kakasi-9e0825a02c7ea5605e968f6208f769f7c49d6860/lib/libkakasi.c", line 744, in kakasi_do [0x65bea3]
E20230506 20:16:24.620041    12 backward.hpp:4199] #0    Source "/typesense/external-Linux/kakasi-9e0825a02c7ea5605e968f6208f769f7c49d6860/lib/libkanjiio.c", line 211, in getpbstr [0x65c624]
Segmentation fault (Address not mapped to object [0x10])
I20230506 20:16:25.094942    12 batched_indexer.cpp:367] Saving currently applying index: 3
E20230506 20:16:25.095103    12 typesense_server.cpp:102] Typesense 0.24.1 is terminating abruptly.
```

## Metadata

**Typesense Version**: 0.24.1

**OS**:
```
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.2 LTS
Release:        22.04
Codename:       jammy
```
```
Server: Docker Engine - Community
 Engine:
  Version:          23.0.5
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.8
  Git commit:       94d3ad6
  Built:            Wed Apr 26 16:19:26 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.20
  GitCommit:        2806fc1057397dbaeefbea0e4e17bddfbd388f38
 runc:
  Version:          1.1.5
  GitCommit:        v1.1.5-0-gf19387a
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Comment by jasonbosco** [comment]
I just tried replicating and can't seem to. Here's the full snippet I used:

```shell
docker run -i -p 8108:8108 -v/tmp/typesense-server-data-1c/:/data typesense/typesense:0.24.1 --data-dir /data --api-key=xyz --listen-port 8108 --enable-cors"
```

```js
const Typesense = require('typesense')

// Create a client
const typesense = new Typesense.Client({
  nodes: [
    {
      host: 'localhost',
      port: '8108',
      protocol: 'http'
    }
  ],
  apiKey: 'xyz',
  numRetries: 3, // A total of 4 tries (1 original try + 3 retries)
  connectionTimeoutSeconds: 120, // Set a longer timeout for large imports
  logLevel: 'debug'
})

async function runExample () {
  try {
    // Delete if the collection already exists from a previous example run
    await typesense.collections('test').delete()
  } catch (error) {
    // do nothing
  }

  try {
    let result
    // create a collection
    result = await typesense.collections().create({ name: "test", fields: [{ name: "japanese", type: "string", locale: "ja" }] });
    console.log(result)

    // create a document
    result = await typesense
      .collections("test")
      .documents()
      .import([{ japanese: "こんにちは" }, { japanese: "またね" }], { action: "upsert" });
    console.log(result)
  } catch (error) {
    console.log(error)
  } finally {
    // Cleanup
    typesense.collections('test').delete()
  }
}

async function timer (seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000))
}

runExample()
```

Output:

```
Request #1683573065073: Performing DELETE request: /collections/test
Request #1683573065073: Nodes Health: Node 0 is Healthy
Request #1683573065073: Updated current node to Node 0
Request #1683573065073: Attempting DELETE request Try #1 to Node 0
Request #1683573065073: Request to Node 0 was made. Response Code was 404.
Request #1683573065097: Performing POST request: /collections
Request #1683573065097: Nodes Health: Node 0 is Healthy
Request #1683573065097: Updated current node to Node 0
Request #1683573065097: Attempting POST request Try #1 to Node 0
Request #1683573065097: Request to Node 0 was made. Response Code was 201.
{
  created_at: 1683573065,
  default_sorting_field: '',
  enable_nested_fields: false,
  fields: [
    {
      facet: false,
      index: true,
      infix: false,
      locale: 'ja',
      name: 'japanese',
      optional: false,
      sort: false,
      type: 'string'
    }
  ],
  name: 'test',
  num_documents: 0,
  symbols_to_index: [],
  token_separators: []
}
Request #1683573065104: Performing POST request: /collections/test/documents/import
Request #1683573065104: Nodes Health: Node 0 is Healthy
Request #1683573065104: Updated current node to Node 0
Request #1683573065104: Attempting POST request Try #1 to Node 0
Request #1683573065104: Request to Node 0 was made. Response Code was 200.
[ { success: true }, { success: true } ]
Request #1683573065110: Performing DELETE request: /collections/test
Request #1683573065110: Nodes Health: Node 0 is Healthy
Request #1683573065110: Updated current node to Node 0
Request #1683573065110: Attempting DELETE request Try #1 to Node 0
Request #1683573065110: Request to Node 0 was made. Response Code was 200.
```

Could you run the snippet in your environment and post a similar output?

**Comment by despenser08** [comment]
Hello, @jasonbosco! I apologize for the delayed response as I was conducting further investigation into the issue. Upon thorough examination, I discovered the cause of the problem by comparing the code I provided as an example with your own code.

It appears that there were no errors in either of the codes. However, upon further analysis using my original document, I identified the specific factor that is leading to the issue.

When you set the field `locale` as `ja` (Japanese) and have duplicate values at the same time, it triggers an error that is unique to this locale. This error does not occur with other locales. To reproduce this, consider the following code snippet that demonstrates how Typesense crashes and becomes unreachable:

```ts
result = await typesense
  .collections("test")
  .documents()
  .import([
    { japanese: "こんにちは" },
    { japanese: "こんにちは" },
    { japanese: "またね" }
  ], { action: "upsert" });
```

Thank you for investigating the issue. If you require any resources from me, please don't hesitate to reach out to me.

**Comment by kishorenc** [comment]
@despenser08 

Thank you for tracking this down. I've a fix in `typesense/typesense:0.25.0.rc29` but it's a X86_64 build -- would you be able to try on a X86 machine?

**Comment by despenser08** [comment]
Hello, unfortunately, I tried to test with Docker on my Windows laptop, and for some reason, document importing did not work. I am unsure whether it is a performance issue with my computer or a general bug. I can see that the request is being sent, but there is no response coming back from the JavaScript SDK. I have attempted this with two different Windows computers using the same code.

**Comment by kishorenc** [comment]
@despenser08 

Hold on, I will create and share an ARM build for you.

**Comment by despenser08** [comment]
> @despenser08
> 
> Hold on, I will create and share an ARM build for you.

@kishorenc, how can I pull the ARM image? I can only see the amd64 version.

**Comment by kishorenc** [comment]
@despenser08 

Please use this ARM image:

```
typesense/typesense:0.25.0.rc30-arm
```

**Comment by despenser08** [comment]
@kishorenc, Thank you for building the ARM image. However, when I attempt to import the document, Typesense does not respond, even though the server is still functioning. I don't think this is a performance issue, as I ran the image on a machine with 4 cores and 24GB of RAM.

**Comment by despenser08** [comment]
I encountered another error while working with my original document:

```
E20230524 13:21:35.460018    33 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460021    30 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460022    35 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460122    24 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460208    31 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460258    35 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460475    35 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460507    35 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460530    35 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460340    42 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460557    40 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460346    23 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:35.460600    35 index.cpp:754] Unhandled Typesense error: at: key not present
E20230524 13:21:36.162384    21 backward.hpp:4199] Stack trace (most recent call last) in thread 21:
E20230524 13:21:36.162411    21 backward.hpp:4199] #16   Object "/usr/lib/aarch64-linux-gnu/ld-linux-aarch64.so.1", at 0xffffffffffffffff, in 
E20230524 13:21:36.162415    21 backward.hpp:4199] #15   Object "/usr/lib/aarch64-linux-gnu/libc.so.6", at 0xffffb2845d1b, in 
E20230524 13:21:36.162420    21 backward.hpp:4199] #14   Object "/usr/lib/aarch64-linux-gnu/libc.so.6", at 0xffffb27dd5c7, in 
E20230524 13:21:36.162421    21 backward.hpp:4199] #13   Object "/opt/typesense-server", at 0x1367cbb, in execute_native_thread_routine
E20230524 13:21:36.162423    21 backward.hpp:4199] #12 | Source "/usr/local/include/c++/10.3.0/thread", line 215, in operator()
E20230524 13:21:36.162425    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/thread", line 271, in _M_invoke<0>
E20230524 13:21:36.162428    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/thread", line 264, in __invoke<ThreadPool::ThreadPool(size_t)::<lambda()> >
E20230524 13:21:36.162431    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 95, in __invoke_impl<void, ThreadPool::ThreadPool(size_t)::<lambda()> >
E20230524 13:21:36.162433    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 60, in operator()
E20230524 13:21:36.162436    21 backward.hpp:4199]     | Source "/typesense/include/threadpool.h", line 57, in operator()
E20230524 13:21:36.162438    21 backward.hpp:4199]       Source "/usr/local/include/c++/10.3.0/future", line 1568, in _M_run [0x49c403]
E20230524 13:21:36.162441    21 backward.hpp:4199] #11 | Source "/usr/local/include/c++/10.3.0/future", line 1435, in _M_set_result
E20230524 13:21:36.162443    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/future", line 412, in call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>()>*, bool*), std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>()>*, bool*>
E20230524 13:21:36.162446    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/mutex", line 729, in __gthread_once
E20230524 13:21:36.162448    21 backward.hpp:4199]       Source "/usr/local/include/c++/10.3.0/aarch64-linux-gnu/bits/gthr-default.h", line 700, in _M_run [0x574a1b]
E20230524 13:21:36.162451    21 backward.hpp:4199] #10   Object "/usr/lib/aarch64-linux-gnu/libc.so.6", at 0xffffb27e25d3, in 
E20230524 13:21:36.162452    21 backward.hpp:4199] #9  | Source "/usr/local/include/c++/10.3.0/future", line 572, in operator()
E20230524 13:21:36.162456    21 backward.hpp:4199]       Source "/usr/local/include/c++/10.3.0/bits/std_function.h", line 622, in _M_do_set [0x499f9b]
E20230524 13:21:36.162459    21 backward.hpp:4199] #8  | Source "/usr/local/include/c++/10.3.0/bits/std_function.h", line 292, in __invoke_r<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<_Fn, _Alloc, _Res(_Args ...)>::_M_run<std::_Bind<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>()>, std::allocator<int>, void, {}>::<lambda()>, void>&>
E20230524 13:21:36.162461    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 115, in __invoke_impl<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<_Fn, _Alloc, _Res(_Args ...)>::_M_run<std::_Bind<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>()>, std::allocator<int>, void, {}>::<lambda()>, void>&>
E20230524 13:21:36.162463    21 backward.hpp:4199]       Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 60, in _M_invoke [0x5a4147]
E20230524 13:21:36.162465    21 backward.hpp:4199] #7  | Source "/usr/local/include/c++/10.3.0/future", line 1373, in operator()
E20230524 13:21:36.162468    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/future", line 1432, in __invoke_r<void, std::_Bind<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>()>&>
E20230524 13:21:36.162470    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 110, in __invoke_impl<void, std::_Bind<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>()>&>
E20230524 13:21:36.162473    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 60, in operator()<>
E20230524 13:21:36.162475    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/functional", line 499, in __call<void>
E20230524 13:21:36.162477    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/functional", line 416, in __invoke<Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>&>
E20230524 13:21:36.162480    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 95, in __invoke_impl<void, Index::batch_memory_index(Index*, std::vector<index_record>&, const string&, const tsl::htrie_map<char, field>&, const string&, const std::vector<char, std::allocator<char> >&, const std::vector<char, std::allocator<char> >&, bool)::<lambda()>&>
E20230524 13:21:36.162482    21 backward.hpp:4199]     | Source "/usr/local/include/c++/10.3.0/bits/invoke.h", line 60, in operator()
E20230524 13:21:36.162484    21 backward.hpp:4199]       Source "/typesense/src/index.cpp", line 699, in operator() [0x5a4043]
E20230524 13:21:36.162487    21 backward.hpp:4199] #6    Source "/typesense/src/index.cpp", line 635, in validate_and_preprocess [0x5a383f]
E20230524 13:21:36.162489    21 backward.hpp:4199] #5    Source "/typesense/src/index.cpp", line 312, in compute_token_offsets_facets [0x5a2ce3]
E20230524 13:21:36.162492    21 backward.hpp:4199] #4    Source "/typesense/src/index.cpp", line 1153, in tokenize_string_with_facets [0x5a1f33]
E20230524 13:21:36.162493    21 backward.hpp:4199] #3    Source "/typesense/src/tokenizer.cpp", line 30, in Tokenizer [0x63726b]
E20230524 13:21:36.162496    21 backward.hpp:4199] #2    Source "/typesense/src/tokenizer.cpp", line 63, in init [0x636f0f]
E20230524 13:21:36.162498    21 backward.hpp:4199] #1    Source "/typesense/external-Linux/kakasi-9e0825a02c7ea5605e968f6208f769f7c49d6860/lib/libkakasi.c", line 744, in kakasi_do [0x65bea3]
E20230524 13:21:36.162501    21 backward.hpp:4199] #0    Source "/typesense/external-Linux/kakasi-9e0825a02c7ea5605e968f6208f769f7c49d6860/lib/libkanjiio.c", line 220, in getpbstr [0x65c668]
Segmentation fault (Address not mapped to object [0x10])
```

**Comment by despenser08** [comment]
As previously mentioned in [this comment](https://github.com/typesense/typesense/issues/1002#issuecomment-1561117326), I was able to successfully import the document. However, I am still perplexed by the fact that my original document continues to generate [this error](https://github.com/typesense/typesense/issues/1002#issuecomment-1561152998).

**Comment by despenser08** [comment]
Sorry, that was from an older version. However, I am experiencing inconsistent behavior where it occasionally displays an "Invalid collection import" error, sometimes it works, and most of the time, Typesense appears to be active as I can observe the peer refreshing process, but the importing process gets stuck.

**Comment by kishorenc** [comment]
@despenser08 

If you are able to share the dataset that causes this issue with me (you can email to contact@typesense.org) I can reproduce the issue and see if there's something else that's going wrong.

**Comment by despenser08** [comment]
@kishorenc, [Here is the dataset I used.](https://s3.camorad.cloud/pub/material.json)

**Comment by kishorenc** [comment]
Thank you, can also please post the collection schema you are using?

**Comment by despenser08** [comment]
@kishorenc

```javascript
fields: [
    { name: 'name_kr', type: 'string', locale: 'ko', optional: false },
    { name: 'name_chs', type: 'string', locale: 'zh', optional: false },
    { name: 'name_cht', type: 'string', locale: 'zh', optional: false },
    { name: 'name_jp', type: 'string', locale: 'ja', optional: false },
    { name: 'name_ru', type: 'string', locale: 'ru', optional: false },
    { name: 'name_th', type: 'string', locale: 'th', optional: false },
    { name: 'desc_kr', type: 'string', locale: 'ko', optional: false },
    { name: 'desc_chs', type: 'string', locale: 'zh', optional: false },
    { name: 'desc_cht', type: 'string', locale: 'zh', optional: false },
    { name: 'desc_jp', type: 'string', locale: 'ja', optional: false },
    { name: 'desc_ru', type: 'string', locale: 'ru', optional: false },
    { name: 'desc_th', type: 'string', locale: 'th', optional: false },
    { name: 'nid', type: 'int32' },
    { name: 'stars', type: 'int32', facet: true, optional: true }
  ],
  default_sorting_field: 'nid'
}
```

**Comment by kishorenc** [comment]
I'm able to reproduce. Looking into it.

**Comment by kishorenc** [comment]
When we index Japanese, we convert all Kanji characters to Hiragana. The library that we use to do that is choking on the following:

"魈"

We've to figure out why that's happening and fix that.

**Comment by despenser08** [comment]
Thank you for taking the time to investigate the issue. Once again, please feel free to reach out to me if there's anything I can do to assist you.

**Comment by kishorenc** [comment]
@despenser08 

I've a question: how important is it that Kanji, Hiragana and Katanaka forms are "normalized" before indexing for Japanese? The problem I am running into is that the library we depend on to do this conversion is really old and unmaintained so it has a few edge cases.  I'm not even sure how accurate it is anymore given how languages evolve. 

If Typesense just indexed Japanese text as it is, would that be fine? This would mean that if there is a word indexed in Kanji, you will not be able to find it via Hiragana writing form.

**Comment by kishorenc** [comment]
I've found a work around. Will polish it up and publish in the next couple of days.

**Comment by despenser08** [comment]
@kishorenc, Thank you for the fix. I'm sorry to tell you that while indexing worked, searching with Japanese did not work for my documents. I'm not sure whether this issue is from my documents or the searching options that I use.

**Comment by kishorenc** [comment]
@despenser08 

Can you please share a query that doesn't work? Also, what you expect to get and what hits you got instead.

**Comment by kishorenc** [comment]
The query string is tokenized based on the locale of the first `query_by` field. We probably need to think about how to tokenize the query separately for each locale independently when multiple languages are used.

**Comment by despenser08** [comment]
Is multiple locale support currently unavailable? Any plans to implement the feature in the future?

**Comment by kishorenc** [comment]
@despenser08 

We do have this on our roadmap. We will work on this shortly after shipping 0.25 (which is on code freeze right now).

**Comment by kishorenc** [comment]
The crash issue is fixed on 0.25.0.
